File: cogs/player.py
import asyncio
import discord
from discord.ext import commands
from utils import logger
import json

from utils import YTDLPWrapper

FFMPEG_OPTIONS = {
    'before_options': '-reconnect 1 -reconnect_streamed 1 -reconnect_delay_max 1000',
    'options': '-vn -reconnect 1 -reconnect_streamed 1 -reconnect_delay_max 1000',
}

class MusicPlayer:
    def __init__(self, url: str, voice_client, interaction: discord.Interaction, resetFunc, server_id, client):
        self.originalUrl: str = url
        self.playlist = [] # { url: string, title: string }[]
        self.playlistMetadata = None # { url: string, title: string }
        self.voice_client: discord.VoiceClient | discord.VoiceProtocol = voice_client
        self.interaction: discord.Interaction = interaction
        self.resetFunc = resetFunc
        self.server_id = server_id
        self.client = client

    async def startLoop(self):
        tmp = YTDLPWrapper().run(["--flat-playlist" ,"-j", self.originalUrl])

        if tmp == "Error" or tmp == "": return self.resetFunc(self.server_id)

        if isinstance(tmp, str):
            tmp = json.loads(tmp)
            self.playlist.append({ "url": tmp["original_url"], "title": tmp["title"] })
        elif isinstance(tmp, list):
            self.playlist = list(map(lambda x: { "url": json.loads(x)["original_url"], "title": json.loads(x)["title"] }, tmp))

        if len(self.playlist) > 1:
            asdads = json.loads(tmp[-1])
            self.playlistMetadata = {
                "url": asdads["playlist_webpage_url"],
                "title": asdads["playlist_title"]
            }

        if self.playlistMetadata:
            embed = discord.Embed(
                title=f":arrow_forward: Playing Playlist: {self.playlistMetadata["title"]}",
                description=self.playlistMetadata["url"],
                color=discord.Color.green()
            )
            await self.interaction.followup.send(embed=embed)

        await self.extractMusic(self.playlist[0]["url"], self.playlist[0]["title"])

    async def extractMusic(self, url: str, title: str):
        try:
            info = await asyncio.to_thread(YTDLPWrapper().run, ["-f", "bestaudio/best", "-g", url])
            print(info)
            if info == "" or info == "Error":
                embed = discord.Embed(
                    title=f":x: Failed Load: {title}",
                    description=url,
                    color=discord.Color.red()
                )
                return await self.interaction.followup.send(embed=embed)
            await self.play(info, title, url)
        except Exception as e:
            logger.error(f"Error: {e}", exc_info=True)
            await self.voice_client.disconnect()
    
    async def detectError(self):
        try:
            self.playlist.pop(0)
            if len(self.playlist) > 0:
                await self.extractMusic(self.playlist[0]["url"], self.playlist[0]["title"])
                return
            self.resetFunc(self.server_id)
            embed = discord.Embed(
                title=f"Ending Playlist",
                color=discord.Color.green()
            )
            await self.interaction.followup.send(embed)
            await self.voice_client.disconnect()
        except Exception as e:
            logger.error(f"Error: {e}", exc_info=True)
            self.resetFunc(self.server_id)
            await self.voice_client.disconnect()
    
    def wrapper(self, error):
        try:
            if error:
                logger.error(f"Error player: {error}")
                return
            self.client.loop.create_task(self.detectError())
        except Exception as e:
            logger.error(f"Error: {e}", exc_info=True)
            
    async def play(self, rawLink: str, title: str, url: str):
        try:
            embed = discord.Embed(
                title=f":arrow_forward: Playing Music: {title}",
                description=url,
                color=discord.Color.green()
            )
            source = discord.FFmpegPCMAudio(rawLink, **FFMPEG_OPTIONS)
            await self.interaction.followup.send(embed=embed)
            self.voice_client.play(source, after=self.wrapper)
        except Exception as e:
            logger.error(f"Error: {e}", exc_info=True)
            await self.voice_client.disconnect()
    
    async def AddContent(self, url, interaction: discord.Interaction):
        tmp = await asyncio.to_thread(YTDLPWrapper().run, ["--flat-playlist" ,"-j", url])

        if tmp == "Error" or tmp == "":
            embed = discord.Embed(
                title=f":x: Failed Adding New Content",
                description=url,
                color=discord.Color.green()
            )
            await interaction.followup.send(embed=embed)
            return
        
        titles = ""

        if isinstance(tmp, str):
            self.playlist.append({ "url": tmp["url"], "title": tmp["title"] })
            titles = f":white_check_mark: Succesfully Adding: {tmp["title"]}"
        elif isinstance(tmp, list):
            self.playlist = self.playlist + list(map(lambda x: { "url": json.loads(x)["url"], "title": json.loads(x)["title"] }, tmp))
            titles = f":white_check_mark: Succesfully Adding Playlist: {json.loads(tmp)[0]["playlist_title"]}"

        embed = discord.Embed(
            title=titles,
            description=url,
            color=discord.Color.green()
        )
        await interaction.followup.send(embed=embed)

class player(commands.Cog):
    def __init__(self, bot):
        self.bot = bot
        self.cache = [] # { ID: MusicPLayer }

    def getCacheFunction(self, guild_id: int):
        try:
            data = set(map(lambda x: x[guild_id] if guild_id in x else None, self.cache))
            return list(filter(lambda x: x is not None, data))
        except Exception as e:
            logger.error(f"Error: {e}", exc_info=True)
    
    def removeFromCache(self, id: int):
        try:
            for i, item in enumerate(self.cache):
                if int(id) in item:
                    self.cache.pop(i)
                    logger.info(f"Cache {id} removed")
        except Exception as e:
            logger.error(f"Error: {e}", exc_info=True)

    @discord.app_commands.command(name="play", description="Play Music")
    async def play(self, interaction: discord.Interaction, url: str):
        await interaction.response.defer(thinking=True)
        try:
            if not interaction.user.voice or not interaction.user.voice.channel:
                await interaction.followup.send('You need be in the voice channel.', ephemeral=True)
                return
            
            voice_client = interaction.guild.voice_client
            channel = interaction.user.voice.channel
            if voice_client is None:
                voice_client = await channel.connect()
            elif voice_client.channel != channel:
                await voice_client.move_to(channel)
            
            if len(self.getCacheFunction(interaction.guild_id)) == 1:
                func = self.getCacheFunction(interaction.guild_id)[0]
                await func.AddContent(url, interaction)
            else:
                tmp = MusicPlayer(url, voice_client, interaction, self.removeFromCache, interaction.guild_id, self.bot)
                await tmp.startLoop()
                self.cache.append(
                    {
                        interaction.guild_id: tmp
                    }
                )

        except Exception as e:
            logger.error(f"Error: {e}", exc_info=True)
            await interaction.followup.send(f"Sorry, {e}")

    @discord.app_commands.command(name="pause", description="Pause Music")
    async def pause(self, interaction: discord.Interaction):
        await interaction.response.defer(thinking=True)
        try:
            if interaction.guild.voice_client:
                if interaction.guild.voice_client.is_paused():
                    interaction.guild.voice_client.resume()
                    await interaction.followup.send("Music is resume :arrow_forward:")
                else:
                    interaction.guild.voice_client.pause()
                    await interaction.followup.send("Music is paused :pause_button:")
            else:
                await interaction.followup.send("No music is playing.", ephemeral=True)
        except Exception as e:
            logger.error(f"Error: {e}")
            await interaction.followup.send(f"Sorry, {e}")

    @discord.app_commands.command(name="stop", description="Stop Music")
    async def stop(self, interaction: discord.Interaction):
        await interaction.response.defer(thinking=True)
        try:
            if interaction.guild.voice_client:
                if interaction.guild.voice_client.is_paused() != True:
                    interaction.guild.voice_client.pause()
                
                self.removeFromCache(interaction.guild_id)
                await interaction.followup.send("I Stopped the music :stop_button:")
            else:
                await interaction.followup.send("No music is playing.", ephemeral=True)
        except Exception as e:
            logger.error(f"Error: {e}")
            await interaction.followup.send("Sorry, i have some problem")

    @discord.app_commands.command(name="leave", description="Leave the voice chat")
    async def leave(self, interaction: discord.Interaction):
        await interaction.response.defer(thinking=True)
        try:
            voice_client = interaction.guild.voice_client
            if voice_client and voice_client.is_connected():
                await voice_client.disconnect()
                self.removeFromCache(interaction.guild_id)
                await interaction.followup.send("I'm Leaving.")
            else:
                await interaction.followup.send("I'm not in the voice chat", ephemeral=True)
        except Exception as e:
            logger.error(f"Error: {e}", exc_info=True)
            self.removeFromCache(interaction.guild_id)
            await interaction.followup.send("Sorry, i have some problem")

    @discord.app_commands.command(name="skip", description="Skip music")
    async def skip(self, interaction: discord.Interaction):
        await interaction.response.defer(thinking=True)        
        try:
            func = self.getCacheFunction(interaction.guild_id)[0]
            if len(func.playlist) > 1:
                if interaction.guild.voice_client.is_playing():
                    interaction.guild.voice_client.stop()
                func.playlist.pop(0)
                embed = discord.Embed(
                    title=f":play_pause: Music Skipped to the next one",
                    description=func.playlist[0],
                    color=discord.Color.green()
                )
                await interaction.followup.send(embed=embed)
            else:
                await interaction.followup.send(":x: I can't skip music :x:")
        except Exception as e:
            logger.error(f"Error: {e}", exc_info=True)
            await interaction.followup.send("Sorry, i have some problem")

    @discord.app_commands.command(name="list", description="List music playlist")
    async def list(self, interaction: discord.Interaction):
        await interaction.response.defer(thinking=True)        
        try:
            func = self.getCacheFunction(interaction.guild_id)[0]
            if len(func.playlist) > 0:
                embed = discord.Embed(
                    title=f":green_book: Music in the playlist: {len(func.playlist)}",
                    color=discord.Color.green()
                )
                await interaction.followup.send(embed=embed)
            else:
                await interaction.followup.send(":x: I'm not actual playing :x:")
        except Exception as e:
            logger.error(f"Error: {e}", exc_info=True)
            await interaction.followup.send("Sorry, i have some problem")

async def setup(bot):
    await bot.add_cog(player(bot))
    logger.info("Player Cog Online")

File: src/commands/player.ts
import { ChildProcessWithoutNullStreams, spawn } from "child_process";
import { cogFormat, MusicTrackFormat } from "../utils/types";
import { createAudioPlayer, createAudioResource, joinVoiceChannel, StreamType, VoiceConnection } from "@discordjs/voice";
import { CacheType, ChatInputCommandInteraction, EmbedBuilder, GuildMember, Message } from "discord.js";
import { runYT_DLP } from "../utils/ytdlp";
import logger from "../utils/logger";
import { formatTimeString, ShuffleList } from "../utils/function";
import { GuildBasedChannel } from "discord.js";

class MusicPlayer {
    ffmpegProcess: ChildProcessWithoutNullStreams | undefined
    playlist: MusicTrackFormat[] = []
    channelID: string = ""
    player = createAudioPlayer()
    connection: VoiceConnection | undefined
    member: GuildMember | undefined
    currentTrack: MusicTrackFormat | undefined
    message: ChatInputCommandInteraction<CacheType> | undefined
    clearFunc = (_: string) => { }
    currentTimeSong = 0

    originalMessage = true
    fetchingMetadata = true

    timer: NodeJS.Timeout | undefined

    constructor(currentURL: string, message: ChatInputCommandInteraction<CacheType>, member: GuildMember, channelID: string, clearFunc: (channelID: string) => void) {
        this.channelID = channelID
        this.clearFunc = clearFunc
        this.initial(message, member, currentURL)
    }

    sendMessage = async (content: string, channel?: Message<true>) => {
        if (this.originalMessage) {
            await this.message?.editReply(content)
            return
        }

        try {
            if (channel) return await channel.edit(content)

            const response: GuildBasedChannel | null = await this.message!.guild!.channels.fetch(this.message!.channel!.id)
            if (!response || response == null) return
            if (!response.isTextBased()) return

            return await response.send(content)
        } catch (error) {
            logger.error(`Failed Fetch Channel ${this.channelID}`)
            return
        }
    }

    extractVideo = async (url: string, cache?: { [key: string]: any }) => {
        this.fetchingMetadata = true
        let response = cache

        const tmpChannel = await this.sendMessage(":arrows_counterclockwise: Fetching Music")
        if (!cache) {
            try {
                response = await runYT_DLP(["-j", url]) as any
            } catch (error) {
                logger.error(`Failed Parse video ${url} ${error}`)
            }
        }

        if (!response || typeof response != "object") {
            this.playlist.shift()

            if (this.playlist.length <= 0) this.selfDestruct()
            return
        }

        let audioUrl: string[] = response["formats"].filter((v: any) => v["resolution"] == "audio only").map((v: any) => v["url"])

        this.currentTrack = {
            url: url == "" ? response["original_url"] : url,
            durration: response["duration"],
            title: response["title"],
            playlistUrl: response["playlist_webpage_url"],
            uploader: response["channel"],
            playlistName: response["playlist"],
            thumbnail: response["thumbnail"]
        }

        audioUrl.reverse()

        const workingUrl = await this.checkWorkingUrl(audioUrl)
        if (!workingUrl) return await this.sendMessage(`:no_entry_sign: Failed Play, ${response["title"]}`)

        this.fetchingMetadata = false
        await this.sendMessage(`:arrow_forward: Playing Music **[${response["title"]}](<${response["original_url"]}>)** (\`${formatTimeString(response["duration"])}\`)`, tmpChannel)
        this.startPlay(workingUrl)
        this.originalMessage = false
    }

    checkWorkingUrl = async (urls: string[]) => {
        for (const url of urls) {
            const ffmpeg = spawn("ffmpeg", [
                "-v", "error",
                "-i", url,
                "-t", "1",
                "-f", "null",
                "-"
            ]);

            const success = await new Promise<boolean>((resolve) => {
                ffmpeg.on("close", (code) => {
                    resolve(code === 0);
                });
            });

            if (success) return url;
        }

        return null;
    }

    AddNewEntries = async (url: string) => {
        if (!this.connection || !this.message) return
        this.fetchingMetadata = true

        const playlistLen = structuredClone(this.playlist.length)

        let playlistLenght = 0

        try {
            this.sendMessage(":arrows_counterclockwise: Fetching Metadata")
            const response: any = await runYT_DLP(["-j", "--flat-playlist", url])

            console.log(response)

            if (typeof response == "string") {
                response.split("\n").forEach((v) => {
                    try {
                        const parsedObject = JSON.parse(v)

                        this.playlist.push({
                            url: parsedObject["url"],
                            durration: parsedObject["duration"],
                            title: parsedObject["title"],
                            playlistUrl: parsedObject["playlist_webpage_url"],
                            playlistName: parsedObject["playlist"],
                            uploader: parsedObject["channel"],
                            thumbnail: parsedObject["thumbnail"]
                        })
                        playlistLenght += 1
                    } catch (error) { }
                })

                if (this.playlist.length <= 0) {
                    this.connection.disconnect()
                    return await this.sendMessage(":no_entry_sign: Failed Fetch Playlist")
                }

                if (playlistLen <= 0) {
                    this.extractVideo(this.playlist[0]["url"])
                    this.currentTrack = this.playlist[0]
                } else {
                    const tmp = this.playlist[this.playlist.length - 1]
                    await this.sendMessage(`:notes: Succesfully Added Playlist **[${tmp.playlistName}](<${tmp.playlistUrl}>)** (\`${playlistLenght}\`)`)
                    this.fetchingMetadata = false
                }

                return
            }

            if (typeof response == "object") {

                this.playlist.push({
                    url: url,
                    durration: response["duration"],
                    title: response["title"],
                    playlistUrl: response["playlist_webpage_url"],
                    playlistName: response["playlist"],
                    uploader: response["channel"],
                    thumbnail: response["thumbnail"]
                })

                if (playlistLen <= 0) {
                    this.extractVideo("", response as any)
                } else {
                    await this.sendMessage(`:musical_note: Succesfully Added **[${response["title"]}](<${response["original_url"]}>)** (\`${formatTimeString(response["duration"])}\`) To The Playlist`)
                    this.fetchingMetadata = false
                }

                return
            }

            this.fetchingMetadata = false

            if (this.playlist.length > 0) return

            await this.sendMessage(":no_entry_sign: Failed Fetch Video")
            this.connection.disconnect()
            return
        } catch (error) {
            this.fetchingMetadata = false
            logger.error("Failed Run URL ", url, error)
            await this.sendMessage(":no_entry_sign: Failed Fetch Video")
            this.connection.disconnect()
        }
    }

    initial = async (message: ChatInputCommandInteraction<CacheType>, member: GuildMember, initialURL: string) => {
        if (!member.voice.channel?.isVoiceBased()) {
            await this.sendMessage("I can't Join to this Text Chat")
            return this.selfDestruct()
        }

        if (!member.voice.channel.joinable) {
            await this.sendMessage("I can't Join to this Voice Chat")
            return this.selfDestruct()
        }

        this.connection = joinVoiceChannel({
            channelId: member.voice.channel!.id,
            guildId: message.guild!.id,
            adapterCreator: message.guild!.voiceAdapterCreator,
            selfDeaf: false
        });

        this.connection.subscribe(this.player)
        this.member = member
        this.message = message

        this.AddNewEntries(initialURL)
    }

    startPlay = async (url: string) => {
        this.currentTimeSong = 0
        if (this.timer) clearInterval(this.timer)
        if (this.ffmpegProcess) this.ffmpegProcess.kill()

        this.ffmpegProcess = spawn("ffmpeg", [
            "-i", url,
            "-f", "s16le",
            "-ar", "48000",
            "-ac", "2",
            "-fflags", "nobuffer",
            "-flags", "low_delay",
            "-reconnect", "1000",
            "-reconnect_streamed", "1000",
            "-reconnect_delay_max", "30",
            "-reconnect_at_eof", "100",
            "-reconnect_on_network_error", "20",
            "-reconnect_on_http_error", "4xx,5xx",
            "-probesize", "32",
            "-analyzeduration", "0",
            "pipe:1"
        ]);

        this.ffmpegProcess.on("close", (err) => {
            console.log("FFMPEG CLOSED EXIT", err)

            if (err != 0) return

            this.playlist.shift()
            if (this.playlist.length <= 0) return this.selfDestruct()

            this.extractVideo(this.playlist[0]["url"])
        })

        const resource = createAudioResource(this.ffmpegProcess.stdout, {
            inputType: StreamType.Raw
        });

        this.timer = setInterval(() => {
            this.currentTimeSong += 1
        }, 1000)
        this.player.play(resource);
    }

    skipSong = async (num: number = 1) => {
        if (this.playlist.length <= num) {
            await this.message?.editReply(":no_entry_sign: Sorry I Can't Skip")
            return
        }
        console.log(num)
        this.playlist = this.playlist.slice(num - 1)
        const current = this.playlist[0]

        if (!current) return this.message?.editReply(":no_entry_sign: Sorry I Can't Skip")

        this.message?.editReply(`:fast_forward: Skipped To **[${current["title"]}](<${current["url"]}>)** (\`${formatTimeString(current["durration"])}\`)`)

        this.extractVideo(this.playlist[0]["url"])
    }

    selfDestruct = () => {
        if (!this.connection) return logger.warn(`MusicPlayer ${this.channelID} Can't Be Destroyed`)
        this.connection?.disconnect()
        this.player.stop(true)
        this.clearFunc(this.channelID)
    }
}

class StreamManagment {
    musicPlayerInstances: MusicPlayer[] = []

    clearObject = (channelID: string) => {
        this.musicPlayerInstances = this.musicPlayerInstances.filter((v) => v["channelID"] != channelID)
    }

    getInstance = async (message: ChatInputCommandInteraction<CacheType>, error: string) => {
        await message.deferReply()

        if (!message.guild) {
            await message.editReply(error);
            return
        }
        const member = await message.guild.members.fetch(message.user.id);

        if (!member) {
            await message.editReply(error);
            return
        }

        if (!member.voice.channel) {
            await message.editReply(":microphone2: You Aren't Connected In Voice Chat")
            return
        }

        const instance = this.musicPlayerInstances.find((v) => v["channelID"] == member.voice.channel!.id)

        if (!instance) {
            await message.editReply(":microphone2: Sorry, I'm not in Voice chat");
            return { instance: undefined, member: member }
        }

        return { instance: instance, member: member }
    }

    connect = async (message: ChatInputCommandInteraction<CacheType>) => {
        const instance = await this.getInstance(message, "Sorry, I can't Leave")
        if (!instance || !instance.member) return

        const url = message.options.getString("url");
        if (!url) return await message.editReply("I Dind't Find Any Url")

        if (!instance.instance) return this.musicPlayerInstances.push(new MusicPlayer(url, message, instance.member, instance.member!.voice!.channel!.id, this.clearObject))

        if (instance.instance.fetchingMetadata) {
            await message.editReply(":warning: I can't add new Music While I'm Fetching other")
            return
        }

        instance.instance.message = message
        instance.instance.AddNewEntries(url)

    }

    disconnect = async (message: ChatInputCommandInteraction<CacheType>) => {
        await message.deferReply()

        const instance = await this.getInstance(message, "Sorry, I can't Leave")
        if (!instance || !instance.instance) return

        instance.instance.selfDestruct()
        await message.editReply("Okay, I'm Leaving")
        this.musicPlayerInstances = this.musicPlayerInstances.filter((v) => v["channelID"] != instance.member.voice.channel!.id)
    }

    pause = async (message: ChatInputCommandInteraction<CacheType>) => {
        const instance = await this.getInstance(message, "Sorry, I can't Pause")
        if (!instance || !instance.instance) return

        if (instance.instance.fetchingMetadata) {
            await message.editReply(":warning: No Music Is Current Play")
            return
        }

        if (instance.instance.player.state.status) {
            instance.instance.player.pause()
            await message.editReply(":pause_button: Music Has Been Pause");
        } else {
            instance.instance.player.unpause()
            await message.editReply(":arrow_forward: Music Has Been UnPause");
        }
    }

    getMusicList = async (message: ChatInputCommandInteraction<CacheType>) => {
        const instance = await this.getInstance(message, "Sorry, I can't Show You The Playlist")
        if (!instance || !instance.instance) return

        let string = `\`\`\`ansi\nCurrent Music Playlist (${instance.instance.playlist.length})\n\n`

        for (let index = 0; index < instance.instance.playlist.length; index++) {
            const v = instance.instance.playlist[index];

            if (string.length >= 1800) {
                string += `... Other (${(instance.instance.playlist.length - 1) - index})`
                break
            }

            const currentPlaying = v["title"] == instance.instance.currentTrack?.title

            const greenText = currentPlaying ? `[32m${v["title"]}[0m` : v["title"]

            string += `${index + 1}. ${greenText} ([36m${formatTimeString(v["durration"])}[0m) ${currentPlaying ? "[1;92mPlaying Now[0m" : ""}\n`
        }

        string += "```"

        await message.editReply(string)
    }

    skipSong = async (message: ChatInputCommandInteraction<CacheType>) => {
        const instance = await this.getInstance(message, "Sorry, I can't Show You The Playlist")
        if (!instance || !instance.instance) return

        if (instance.instance.fetchingMetadata) {
            await message.editReply(":warning: I Can't Change The Song When Other Song Is Loading")
            return
        }

        const number = message.options.getNumber("position")

        instance.instance.message = message
        instance.instance.skipSong(number ? number : 1)
    }

    shufflePlaylist = async (message: ChatInputCommandInteraction<CacheType>) => {
        const instance = await this.getInstance(message, "Sorry, I can't Shuffle The Playlist")
        if (!instance || !instance.instance) return

        if (instance.instance.fetchingMetadata) {
            await message.editReply(":warning: I can't Shuffle Playlist When Song Is Fetching")
            return
        }

        if (instance.instance.playlist.length <= 1) {
            await message.editReply(":warning: I can't Shuffle Playlist because is too small")
            return
        }

        let tmpPlaylist = structuredClone(instance.instance.playlist)
        const firstElement = tmpPlaylist[0]
        tmpPlaylist = tmpPlaylist.slice(1)

        tmpPlaylist = ShuffleList(tmpPlaylist)
        tmpPlaylist.unshift(firstElement)

        instance.instance.playlist = tmpPlaylist

        await message.editReply(":white_check_mark: Succesfully Shuffle The Playlist")
    }

    MusicStatus = async (message: ChatInputCommandInteraction<CacheType>) => {
        const instance = await this.getInstance(message, "Sorry, I Can't Show Status of Current Music")
        if (!instance || !instance.instance) return

        if (instance.instance.fetchingMetadata) {
            await message.editReply(":warning: I Can't Give Music Status when is loading")
            return
        }

        const isPlaying = instance.instance.player.state.status
        const current = instance.instance.currentTrack
        const currentTime = instance.instance.currentTimeSong

        if (!current) {
            await message.editReply(":warning: No Music Is Current Play")
            return
        }

        await message.editReply(`${isPlaying ? ":arrow_forward:" : ":pause_button:"} Music: **[${current["title"]}](<${current["url"]}>)** (\`${formatTimeString(currentTime)}/${formatTimeString(current["durration"])}\`)`)
    }
}

export const PlayerManager = new StreamManagment

export default {
    name: "player",
    commands: [],
    slashCommands: [{
        name: "playsong",
        execute: PlayerManager.connect,
        description: "Play Song In Bot",
        options: [{ name: "url", required: true, type: 3 }]
    }, {
        name: "leavevoice",
        execute: PlayerManager.disconnect,
        description: "Leaving Voice Channel",
    }, {
        name: "pausemusic",
        execute: PlayerManager.pause,
        description: "Pause Music In Voice Chat",
    }, {
        name: "playlistnow",
        execute: PlayerManager.getMusicList,
        description: "Get Full Music Playlist",
    }, {
        name: "skipmusic",
        execute: PlayerManager.skipSong,
        description: "Skip Music",
        options: [{ name: "position", required: false, type: 10 }]
    },{
        name: "addsongs",
        execute: PlayerManager.connect,
        description: "Add New Song Or Playlist",
        options: [{ name: "url", required: true, type: 3 }]
    },{
        name: "musicstatus",
        execute: PlayerManager.MusicStatus,
        description: "Check Music Status",
    },{
        name: "shuffleplaylist",
        execute: PlayerManager.shufflePlaylist,
        description: "Shuffle Current Playlist",
    }]
} as cogFormat